tvb_library/tvb/datasets/tvb_data.py

```diff
@@ -0,0 +1,190 @@
+#
+#
+#   CITATION:
+# When using The Virtual Brain for scientific publications, please cite it as explained here:
+# https://www.thevirtualbrain.org/tvb/zwei/neuroscience-publications
+#
+#
+
+"""
+.. moduleauthor:: Abhijit Deo <f20190041@goa.bits-pilani.ac.in>
+"""
+
+import os
+import requests
+import json
+import pooch
+from pathlib import Path
+from zipfile import ZipFile
+import shutil
+
+from .base import BaseDataset
+from .zenodo import Zenodo, Record, BASE_URL
+
+class TVBZenodoDataset(BaseDataset):
+
+    CONCEPTID = "3417206"
+    
+    def __init__(self, version= "2.7", extract_dir = None):
+        """
+        Constructor for TVB_Data class 
+
+        parameters
+        -----------
+
+        version: str
+              - Version number of the dataset, Default value is 2.7
+
+        """
+        super().__init__(version, extract_dir)
+        self.cached_dir = self.extract_dir / ".cache" 
+        self.cached_file = self.cached_dir / "tvb_cached_responses.txt"
+
+        if  not self.cached_dir.is_dir():
+            self.cached_dir.mkdir(parents=True)
+
+        try:
+            self.recid = self.read_cached_response()[version]['conceptrecid']
+            
+        except :
+            self.log.warning(f"Failed to read data from cached response.")
+            self.recid = Zenodo().get_versions_info(self.CONCEPTID)[version]            
+            self.update_cached_response()
+        
+
+        #TODO add logging errors method by catching the exact exceptions. 
+        self.rec = Record(self.read_cached_response()[self.version])
+
+    def download(self, path=None):
+        """
+        Downloads the dataset to `path`
+        """
+        self.rec.download(path)
+
+    def _fetch_data(self, file_name):        
+        """
+        Fetches the data 
+
+        parameters:
+        -----------
+        file_name: str
+                - Name of the file from the downloaded zip file to fetch. 
+        extract_dir: str
+                - Path where you want to extract the archive. If Path is None, dataset is extracted according to the tvb profile configuration 
+
+
+        returns: Pathlib.Path
+            path of the file which was extracted
+        """
+        # TODO: extract dir needs better description.
+
+
+        extract_dir = self.extract_dir 
+        download_dir = self.cached_dir / "TVB_Data"
+
+        try:
+            file_path = self.rec.file_loc['tvb_data.zip']
+        except:
+            self.download(path = download_dir)
+            file_path = self.rec.file_loc['tvb_data.zip']
+
+        with ZipFile(file_path) as zf:
+            file_names_in_zip = zf.namelist()
+        zf.close()
+
+        file_name = file_name.strip()
+
+
+        file_names_in_zip = {str(Path(i).name): i for i in file_names_in_zip}
+        if extract_dir==None:
+            ZipFile(file_path).extract(file_names_in_zip[file_name])
+
+        ZipFile(file_path).extract(file_names_in_zip[file_name], path = extract_dir)
```

Comment:
What happens if the same file is fetched twice ?

Reply:
We then again go to our tvb_data.zip, get the file from there, and replace the older existing one. 

I do understand that we might want to avoid this. 

What we can do is, we do know where the file is going to be unzipped (as in the path of the file).  So if the file already exists on that path, then return the path; we don't make any changes to the file system. 


On a side note, one case might be that the file already existing at that path might contain some different data (i.e. tampered file; very rare case indeed); then the current approach is safer.

Reply:
We have the method to delete the repo, in case we discover that is tampered. I think that is enough.
Instead downloading the full zip again is costly (at least in my flows ;-) )